be/src/vec/exec/executor/listenable_future.h

```diff
@@ -0,0 +1,219 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <condition_variable>
+#include <exception>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+#include "common/status.h"
+#include "glog/logging.h"
+
+namespace doris {
+namespace vectorized {
+
+struct Void {
+    bool operator==(const Void&) const { return true; }
+    bool operator!=(const Void&) const { return false; }
+};
+
+template <typename T>
+class SharedListenableFuture;
+
+template <typename T>
+class ListenableFuture {
+public:
+    using Callback = std::function<void(const T&, const doris::Status&)>;
+
+    ListenableFuture(const ListenableFuture&) = delete;
+    ListenableFuture& operator=(const ListenableFuture&) = delete;
+
+    ListenableFuture(ListenableFuture&& other) noexcept
+            : _ready(other._ready),
+              _value(std::move(other._value)),
+              _status(std::move(other._status)),
+              _callbacks(std::move(other._callbacks)) {
+        other._ready = false;
+    }
+
+    ListenableFuture& operator=(ListenableFuture&& other) noexcept {
+        if (this != &other) {
+            std::lock_guard<std::mutex> lock(_mutex);
+            std::lock_guard<std::mutex> other_lock(other._mutex);
+            _ready = other._ready;
+            _value = std::move(other._value);
+            _status = std::move(other._status);
+            _callbacks = std::move(other._callbacks);
+            other._ready = false;
+        }
+        return *this;
+    }
+
+    ListenableFuture() : _ready(false) {}
+
+    void set_value(const T& value) { _execute(value, doris::Status::OK()); }
+
+    void set_error(const doris::Status& status) { _execute({}, status); }
+
+    void add_callback(Callback cb) {
+        std::lock_guard<std::mutex> lock(_mutex);
+        if (_ready) {
+            cb(_value, _status);
```

Comment:
Do the callback within lock may impact the performance?
Just make sure the callback itself is lightweight

Reply:
Good catch.